src/pygeoprocessing/geoprocessing.py

```diff
@@ -2609,20 +2671,44 @@ def warp_raster(
         callback_data=[target_raster_path])
 
     if vector_mask_options:
-        # Make sure the raster creation options passed to ``mask_raster``
-        # reflect any metadata updates
-        updated_raster_driver_creation_tuple = (
-            raster_driver_creation_tuple[0], tuple(raster_creation_options))
-        # there was a cutline vector, so mask it out now, otherwise target
-        # is already the result.
-        mask_raster(
-            (warped_raster_path, 1), vector_mask_options['mask_vector_path'],
-            target_raster_path,
-            mask_layer_id=mask_layer_id,
-            where_clause=mask_vector_where_filter,
-            target_mask_value=None, working_dir=temp_working_dir,
-            all_touched=False,
-            raster_driver_creation_tuple=updated_raster_driver_creation_tuple)
+        if 'mask_raster_path' in vector_mask_options:
+            # If the user provided a mask raster, use that directly; assume
+            # it's been rasterized correctly.
+            source_raster_info = get_raster_info(warped_raster_path)
+            source_nodata = source_raster_info['nodata'][0]
+
+            def _mask_values(array, mask):
+                output = numpy.full(array.shape, source_nodata)
+                valid_mask = (
+                    mask == 1 & ~array_equals_nodata(array, source_nodata))
+                output[valid_mask] = array[valid_mask]
+                return output
+
+            raster_calculator(
+                [(warped_raster_path, 1),
+                 (vector_mask_options['mask_raster_path'], 1)],
+                _mask_values, target_raster_path,
+                source_raster_info['datatype'], source_nodata)
+        else:
+            # If the user did not provide a mask raster, we need to create one.
+            # mask_raster() does the rasterization for us.
```

Comment:
I'm a little confused by this comment, maybe it's a leftover? If we're here then we're masking the warped target by a vector, right? No need to create a mask raster first, just use `mask_raster` as you have done?

Reply:
This comment could be improved, so I tried to do so in e5c8557f.

If we've gotten to this point, the user provided a mask vector, which means that we need to 1) create a new raster from base for the rasterization, 2) rasterize the vector and 3) use the rasterized vector mask to determine which pixels are masked out.  `mask_raster` handles all 3 of these steps for us.

Does that help clarify things?